BUG: Refcount fixes

seberg opened the pull request:
Fixes reference count bugs all over the place. Some are a bit more involved, a few small ones may be fly-by that are not excersized by the tests (using debugging python and `pytest-leaks`).

---

At the time of opening it is still a bit of a WIP with more commits coming, but those should not touch the others.

seberg committed:
BUG: Fix leaking of object descriptor on dtype discovery of `None`

seberg committed:
BUG: Fix reference count leak in string discovery code path

seberg committed:
BUG: Fix two more datetime related reference count bugs

One is leaking only the python integer 1, the other a descriptor.

seberg committed:
BUG: Fix bugs in vectorized string functions error paths

Fixes reference count issues (a leak) in
TestVecString::test_non_existent_method and
TestVecString::test_non_string_array

seberg committed:
BUG: Fix Reference count leak in in128 C-level tests

seberg committed:
BUG: Fix refcount leak in advanced indexing error path

seberg committed:
BUG,MAINT: Refactor indexing dtypes handling in mapiter constructor

Using borrowed references for the dtypes seems much easier to get
the reference counts right.

seberg committed:
BUG: Reference leaks in deprecated datetime with timezone coercion

seberg committed:
BUG: Refcount issue in nonzero 0-d path

seberg committed:
BUG: Fix ufunc.at dtype refcount handling in resolver

seberg committed:
BUG: Fix reference count in nditer python op_axes setup

seberg committed:
BUG: Reference leak in writebackifcopy non-array error path

seberg commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -500,10 +501,11 @@ array_dealloc(PyArrayObject *self)
+             * prevent reaching 0 twice and thus recursing into dealloc.
+             * Avoid INCREF for now to avoid increasing sys.gettotalrefcount.
+             */
+            PyArray_REFCOUNT(self) += 1;
```
This is a necessary evil to avoid changing the total reference count, and it should work to prevent the array reaching a reference count of 0 twice. I do not like the hidden struct access, but we have it in a different code path already, so my guess it is just as well (until it breaks).

seberg replied:
Ah, cannot use that macro on PyPy, but I guess the strcut access probably works...

mattip replied:
That macro is an evil hack. Python is moving away from direct access, this will break if CPython ever moves forward with its garbage collecting. We should remove it. Can't you just add a DECREF later?

seberg replied:
As discussed, I do not know. The first two paths we can maybe avoid easily, the last one maybe an ifdef can be done, or we just wait until python actually breaks us and fix it up then. In Principle once the refcount reaches 0, it should probably never increase again, but we need to iterate the array and that uses so much machinery which holds on to references, that it might be annoying to fix for real.

eric-wieser replied:
Can we split out these scary-looking `PyArray_REFCOUNT` changes from the obviously correct ones?

seberg replied:
Will split it off. As a random remark: I just saw that cython does the same thing (i.e. modify reference count without using `INCREF`.

seberg committed:
BUG: Add missing DECREF in ufunc with dtype type resolution

seberg committed:
BUG: Reference count bugs in io functions

seberg committed:
BUG: Reference count loss in as_c_array C-side tests

seberg committed:
BUG: Leaking of method in object method based ufunc loops

The function could possibly be optimized nicely by checking
that the method will be the same because the types are identical.
OTOH, that may not be correct in dynamically adapted types.

seberg commented:
One issue with these fixes is that they do expose other bugs where references are lost, which then can cause segfaults in production. Some things could probably go into 1.17 in principle, but the whole set may be better for later...

seberg committed:
TST: Clean up cyclic ctypes reference to simplify leak debugging

If we do not clean it up in this test, the collection can take
so long that it affects the following tests when running with
reference leak debugging tools (pytest-leaks).

seberg committed:
BUG: Reference count of raw data for 0-sized unpickling

seberg commented:
Hmmm, I got an email that the CI is failing. Looks like it is raising the deprecation warning during test gathering

seberg committed:
TST: Mark tests which must leak references

It really may be time to remove this, considering how annoying this
is, may want to think about getting rid of that behaviour.

Could probably write a multiearraytests C-side function instead.

seberg committed:
BUG: Fix refcount leaks in arr.getield and setfield error paths

seberg committed:
BUG: Fix refcount issue in wheremask

This fix is not the prettiest, since it relies on the wheremask
to be formatted to boolean beforehand correctly.

seberg committed:
TST: Reset StringConvert after mutating it in tests

Or maybe just mark the test?

eric-wieser commented on numpy/lib/tests/test__iotools.py:
```diff
@@ -212,6 +213,7 @@ def test_upgrademapper(self):
         assert_equal(test, date(2009, 1, 1))
         test = convert('')
         assert_equal(test, date(2000, 1, 1))
+        StringConverter._mapper = _original_mapper
```
This should be in a `try` / `finally` , in case an assertion fails

eric-wieser replied:
(is this change related to the patch as a whole?

seberg replied:
This keeps adding to `_mapper` which is a list and thus grows. So it trips `pytest-leaks` because the total reference count increases each time the test runs.

eric-wieser replied:
Got it, thanks

eric-wieser commented on numpy/core/src/multiarray/_multiarray_tests.c.src:
```diff
@@ -928,6 +928,7 @@ test_as_c_array(PyObject *NPY_UNUSED(self), PyObject *args)
 
     num_dims = PyArray_NDIM(array_obj);
     descr = PyArray_DESCR(array_obj);
+    Py_INCREF(descr);
```
nit: add a comment saying `PyArray_AsCArray steals a reference to this`

seberg committed:
BUG: Fix missing static for npy_cache_import in umath funcs.inc.src

eric-wieser commented on numpy/core/src/multiarray/common.c:
```diff
@@ -214,6 +213,13 @@ PyArray_DTypeFromObjectHelper(PyObject *obj, int maxdims,
+            /*
+             * For this branch, _array_find_python_scalar is only good for
+             * finding that it is a scalar.
+             */
```
I don't really understand this comment. Is the point you're making that the input dtype is completely ignored?

seberg replied:
Yeah, that is better.

eric-wieser commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -2236,14 +2236,17 @@ PyArray_Nonzero(PyArrayObject *self)
         static npy_intp const zero_dim_shape[1] = {1};
         static npy_intp const zero_dim_strides[1] = {0};
 
+        Py_INCREF(PyArray_DESCR(self));
```
nit: comment about reference being stolen

eric-wieser commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -2236,14 +2236,17 @@ PyArray_Nonzero(PyArrayObject *self)
-        return PyArray_Nonzero(self_1d);
+        ret_tuple = PyArray_Nonzero(self_1d);
+        Py_DECREF(self_1d);
+        return ret_tuple;
```
Oops, this was recent.

seberg replied:
I could split off the commit to merge quicker and make sure it is in 1.17.

eric-wieser replied:
Probably not a big deal, this path is unlikely to ever be hit anyway (hence my attempt to deprecate it)

eric-wieser commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -2616,7 +2617,8 @@ PyArray_MapIterNew(npy_index_info *indices , int index_num, int index_type,
 
     PyArrayObject *index_arrays[NPY_MAXDIMS];
-    PyArray_Descr *dtypes[NPY_MAXDIMS];
+    PyArray_Descr *intp_descr = PyArray_DescrFromType(NPY_INTP);
```
Missing null check?

seberg replied:
It actually cannot fail for these builtin types, since it is just a table lookup + incref. Of course, maybe that is an implementation detail (it probably was used sloppy around that area more often, but not sure).

eric-wieser replied:
Perhaps just add an `assert` then

seberg replied:
I am adding the error check, but I am wondering if it may make sense to give the no-error guarantee (for our own numeric dtypes).

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -371,6 +371,7 @@ PyArray_GetField(PyArrayObject *self, PyArray_Descr *typed, int offset)
         safe = PyObject_CallFunction(checkfunc, "OOi", PyArray_DESCR(self),
                                      typed, offset);
         if (safe == NULL) {
+            Py_DECREF(typed);
```
Needed above on line 366 too.

seberg replied:
whoops, I think SetField I did not forget that, will add that right away.

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -582,14 +587,13 @@ array_tofile(PyArrayObject *self, PyObject *args, PyObject *kwds)
         return NULL;
     }
     if (PyBytes_Check(file) || PyUnicode_Check(file)) {
-        file = npy_PyFile_OpenFile(file, "wb");
```
This looks wrong to me - `file` is a borrowed reference, but this now decrefs it.

eric-wieser replied:
nevermind

seberg replied:
File is only borrowed until it is replaced by `NpyPath_PathlikeToFspath`

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -5693,18 +5693,13 @@ ufunc_at(PyUFuncObject *ufunc, PyObject *args)
     operands[0] = op1_array;
     if (op2_array != NULL) {
-        dtypes[1] = PyArray_DESCR(op2_array);
-        dtypes[2] = dtypes[0];
```
Why are these no longer assigned to?

seberg replied:
Because they should be filled in by the TypeResolver in any case, initializing it doesn't really do anything.

seberg commented:
This should be actually settling down a little. f2py and distutils is still leaky as hell. And `ufunc.at` leaks in the error path (cleaning that up might be just an extra `iternext` call, but it is pretty broken, and I would have to think more about it).

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -7160,6 +7162,13 @@ def make_ctype(shape, scalar_type):
+        del c_u8_33d, m
+        for i in range(33):
+            if gc.collect() == 0:
+                break
```
Ideally this would go in a `try` / `finally` too

eric-wieser reviewed (commented):
Generally looks great, nice work! Seems you updated while I was commenting though

seberg commented:
Sorry about the live update @eric-wieser, I did not force push, so they should be at the end (not sure how much got added while you were checking things)

eric-wieser commented:
No worries, at this point there are enough leak fixes here that even if you do manage to introduce a new one it's still worth merging.

seberg commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1101,8 +1099,8 @@ PyArray_NewFromDescr_int(
-        if (is_empty) {
-            nbytes = descr->elsize;
+        if (nbytes == 0) {
+            nbytes = descr->elsize ? descr->elsize : 1;
```
I guess this should be split of as well? Since it is not refcount related, just due to partial test suit runs, it blew up (I guess because of the caching).

eric-wieser replied:
I don't care that much, but if you want to shoot off a quick CR fixing this one, I'll approve it

seberg committed:
BUG: Fix refcount in multiternew error path

seberg committed:
BUG: Fix npy_ObjectGCD and LCM reference counting

seberg committed:
BUG: Decrement already wrapped outputs on failure

This is not super pretty, but not sure how to do it much better
right now.

seberg commented:
Still have to address the comments, but ran a full test run over night. There are two flaky tests, and the `ufunc.at` issue (on error within object loop). Other than that, the only remaining issues are in distutils (my guess is probably just deep cycles/flak?) and f2py, which I will not touch here, got an old attempt after all.
So this really should have pretty much settled.

seberg committed:
FIXUP: iotools StringConverter._mapper

seberg committed:
FIXUP: as_c_array tests

seberg committed:
FIXUP: String/dtype discovery comment

seberg committed:
FIXUP: Nonzero comment

seberg committed:
FIXUP: mapiternew -- check error return of dtype creation

In principle, at least at this time, this is not possible

seberg commented:
Except that very last comment went through things (rebased the split off things out). I did not squash yet. I can rebase the fixups into their specific commits, or we can squash the whole lot.

seberg commented on numpy/conftest.py:
```diff
@@ -16,6 +16,8 @@
     config.addinivalue_line("markers",
         "valgrind_error: Tests that are known to error under valgrind.")
+    config.addinivalue_line("markers",
+        "leaks_references: Tests that are known to leak references.")
```
I do not mind if we add them or not. I did not actually use them yet, so a comment in the test will do (or the correct magic to actually manage to create such arrays, I am starting to feel we have been always supporting something that is exceedingly difficult to achieve.
